lib/membrane_rtmp_plugin/rtmp/sink/sink.ex

```diff
@@ -0,0 +1,142 @@
+defmodule Membrane.RTMP.Sink do
+  @moduledoc """
+  Membrane element being client-side of RTMP streams
+
+  Implementation based on FFmpeg.
+  """
+  use Membrane.Sink
+  alias __MODULE__.Native
+  alias Membrane.{AAC, MP4}
+  require Membrane.Logger
+
+  def_input_pad :audio,
+    availability: :always,
+    caps: AAC,
+    mode: :pull,
+    demand_unit: :buffers
+
+  def_input_pad :video,
+    availability: :always,
+    caps: MP4.Payload,
+    mode: :pull,
+    demand_unit: :buffers
+
+  def_options rtmp_url: [
+                type: :string,
+                spec: String.t(),
+                description: "URL address of an RTMP/RTMPS stream destination (e.g. youtube)."
+              ]
+
+  @impl true
+  def handle_init(options) do
+    {:ok, Map.from_struct(options)}
+  end
+
+  @impl true
+  def handle_prepared_to_playing(_ctx, state) do
+    case Native.init_connection(state.rtmp_url) do
+      {:ok, native_state} ->
+        Membrane.Logger.debug("Correctly initialized connection with: #{state.rtmp_url}")
+        state = Map.put(state, :native_state, native_state)
+        {:ok, state}
+
+      {:error, reason} ->
+        raise("#{reason}")
+    end
+  end
+
+  @impl true
+  def handle_playing_to_prepared(_ctx, state) do
+    Native.close_connection(state.native_state)
+    state = Map.delete(state, :native_state)
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_caps(:video, %MP4.Payload{} = caps, ctx, state) do
+    case Native.init_video_stream(state.native_state, caps.width, caps.height, caps.content.avcc) do
+      {:ok, ready, native_state} ->
+        Membrane.Logger.debug("Correctly initialized video stream.")
+        state = Map.put(state, :native_state, native_state)
+
+        if ready do
+          write_header(ctx, state)
+        else
+          {:ok, state}
+        end
```

Comment:
Actually it is more complex. Demanding frames and attempting to send them before writing a header will result in segmentation fault so I need to wait for both caps before demanding buffers from pads. This is what I use `ready` variable for. My suggestion is that I will remove writing header from elixir but I will still return `ready` variable from functions responsible for initializing audio and video streams.

Reply:
I think I would just raise if any of the caps is `nil` and `handle_write/4` is called. What you are doing here is nice, but it uses some magic and the fact that you need these caps to be sent is not clear. Also, some elements, like H264 Parser or AAC Parser need to get a demand before they will generate any caps, because they rely on byte stream data. This Sink only works because Buffer between elements demands some data just to fill itself.

- Demand as always in `handle_prepared_to_playing`
- Raise if `handle_write` is called when any of the caps is missing. You can also check it in native code if you prefer, it doesn't matter to me
- If you do the above, you can just `if(ready) write_header` in native code wherever you calculate `ready` and remove `write_header` from the spec - it will actually be less complex, especially on Elixir side

Reply:
If I demand from both pads in `handle_prepared_to_playing` I may receive first video frame after receiving H264 caps, but before AAC caps. I spotted this case after getting segmentation fault from time to time at irregular rate because I was attempting to write frame prior to sending header. If I implement the steps above it will mean that I allow it to happen but return an informative error instead of segmentation fault. Do we accept this? Let me know if I understand you correctly.

Reply:
No, we do not accept any failure that is caused by race condition in our code. Neither do we accept any magic. What you need to do here is buffer that problematic frame and send it ASAP. If you demand correctly, I think that this will only be one frame, if there is more, you can probably raise, but I'm not a 100% sure, think about it.